### PR TITLE
Add Button component

### PR DESCRIPTION
**Pull request by miguel-s**
Adds the `<Button />` component.

Based on Phoenix's `button/2` https://hexdocs.pm/phoenix_html/Phoenix.HTML.Link.html#button/2.

### CODE REVIEW

File: lib/surface/components/button.ex [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+  @doc "Class or classes to apply to the button"
+  prop class, :css_class
+
+  @doc """
+  The label for the generated `<button>` element, if no content (default slot) is provided.
+  """
+  prop label, :string
+
+  @doc "Triggered on click"
+  prop click, :event
```
Comment (Malian) [inline lib/surface/components/button.ex:50]: Shouldn't we add the others usual events like the ones we find in the `Surface.Components.Form.Submit` component?
```elixir
      @doc "Triggered when the component loses focus"
      prop blur, :event

      @doc "Triggered when the component receives focus"
      prop focus, :event

      @doc "Triggered when the component receives click"
      prop capture_click, :event

      @doc "Triggered when a button on the keyboard is pressed"
      prop keydown, :event

      @doc "Triggered when a button on the keyboard is released"
      prop keyup, :event
```
Moreover, WDYT of renaming `click` to `capture_click` to be consistent in the whole codebase? In that case, we will also modify the `click` event in the `Link` component.

File: lib/surface/components/link.ex [hunk 2]
```diff
@@ -2,8 +2,7 @@ defmodule Surface.Components.Link do
   @moduledoc """
   Generates a link to the given URL.
 
-  Provides similar capabilities to Phoenix's built-in `link/2`
-  function.
+  Provides similar capabilities to Phoenix's built-in `link/2` function.
```
Comment (Malian) [inline lib/surface/components/link.ex:5]:
```suggestion
  Provides similar capabilities to Phoenix's built-in `button/2` function.
```

File: test/components/button_test.exs [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+
+    test "button with invalid args" do
+      msg = """
+      unsupported scheme given to <Button />. In case you want to link to an
+      unknown or unsafe scheme, such as javascript, use a tuple: {:javascript, rest}
+      """
+
+      assert_raise ArgumentError, msg, fn ->
+        render_surface(do: ~H[<Button label="foo" to="javascript:alert(1)" method={{ :get }} />])
+        |> IO.inspect()
```
Comment (Malian) [inline test/components/button_test.exs:210]:
```suggestion
```

**Review by Malian — COMMENTED** [review]
Thanks @miguel-s 

Except the few comments below, the code looks good to me. I am just wondering if we could extract some of the code to a private API and share it between the `Link` and the `Button` components as 95% of the code is the same.

File: test/components/button_test.exs (hunk 3 above)
Comment (miguel-s) [reply test/components/button_test.exs]: Good catch!

File: lib/surface/components/button.ex [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+    "rtsp:",
+    "svn:",
+    "tel:",
+    "fax:",
+    "xmpp:"
+  ]
+
+  def update(assigns, socket) do
+    unless assigns[:default] || assigns[:label] || Keyword.get(assigns.opts, :label) do
+      raise ArgumentError, "<Link /> requires a label prop or contents in the default slot"
```
[the remark this answers is not in the document]
Comment (miguel-s) [reply lib/surface/components/button.ex]: Good catch!

File: lib/surface/components/link.ex (hunk 2 above)
Comment (miguel-s) [reply lib/surface/components/link.ex:5]: Wrong file or am I not seeing something?

File: lib/surface/components/button.ex (hunk 1 above)
Comment (miguel-s) [reply lib/surface/components/button.ex:50]: Yes makes sense, somehow missed these changes.

It seems that in `Surface.Components.Form.Submit` these events are defined, but never actualy used. In `Surface.Components.Form.Reset` they're are defined and used. Is there a reason for this or just an oversight?

**Comment by miguel-s** [comment]
> I am just wondering if we could extract some of the code to a private API and share it between the `Link` and the `Button` components as 95% of the code is the same.

Yes planning on doing just that :)

Waiting on some changes that @msaraiva is planning for Phoenix that will allow us to also clean up all the `opts_to_attrs` logic.

File: lib/surface/components/link.ex (hunk 2 above)
Comment (Malian) [reply lib/surface/components/link.ex:5]: Forget this suggestion. I thought I deleted it.

File: lib/surface/components/button.ex (hunk 1 above)
Comment (msaraiva) [reply lib/surface/components/button.ex:50]: > It seems that in `Surface.Components.Form.Submit` these events are defined, but never actualy used. In `Surface.Components.Form.Reset` they're are defined and used. Is there a reason for this or just an oversight?

Probably an oversight.

**Comment by msaraiva** [comment]
> Waiting on some changes that @msaraiva is planning for Phoenix that will allow us to also clean up all the `opts_to_attrs` logic.

@miguel-s those changes were applied to `phoenix_html` and are already on `master`, however, we have to wait for a new phoenix release that includes the upcoming version of `phoenix_html`. The new function is `Phoenix.HTML.Tag.attributes_escape/1`. More details can be found on the PR: https://github.com/phoenixframework/phoenix_html/pull/331.

**Review by Malian — APPROVED** [review]

**Review by msaraiva — APPROVED** [review]

**Pull request merged**